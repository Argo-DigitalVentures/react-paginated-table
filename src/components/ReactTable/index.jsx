import React, { PureComponent, PropTypes } from 'react';
import find from 'lodash.find';

import TableComponent from '../Table';
import TheadComponent from '../Thead';
import TbodyComponent from '../Tbody';
import TrComponent from '../Tr';
import ThComponent from '../Th';
import TdComponent from '../Td';
import PaginationComponent from '../Pagination';
import PreviousComponent from '../Previous';
import NextComponent from '../Next';
import LoadingComponent from '../Loading';

import {
  dataType,
  columnType,
  sortType,
  styleType,
  classNameType,
} from '../../propTypes';

class ReactTable extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      sorts: [...props.initialSorts],
      pageSize: props.pageSize,
      page: 1,
    };
    this.changeSort = ::this.changeSort;
    this.setPageSize = ::this.setPageSize;
    this.jumpToPage = ::this.jumpToPage;
    this.incrementPage = ::this.incrementPage;
    this.decrementPage = ::this.decrementPage;
  }

  componentWillMount() {
    this.onChange();
  }

  onChange() {
    this.props.onChange({
      ...this.state,
      count: this.props.count,
    });
  }

  setSort(sorts) {
    this.setState({ sorts }, this.onChange);
  }

  setPageSize(pageSize) {
    this.setState({ pageSize, page: 1 }, this.onChange);
  }

  changeSort(columnId, additive) {
    const sort = find(this.state.sorts, sort => sort.columnId === columnId);
    const sorts = additive ? this.state.sorts.filter(sort => sort.columnId !== columnId) : [];
    const newSort = sort && sort.direction === 'DESC' ? [] : [{
      columnId,
      direction: !sort ? 'ASC' : 'DESC',
    }];
    const newSorts = [...sorts, ...newSort];
    this.setSort(newSorts);
  }

  jumpToPage(page) {
    this.setState({ page }, this.onChange);
  }

  incrementPage() {
    this.jumpToPage(this.state.page + 1);
  }

  decrementPage() {
    this.jumpToPage(this.state.page - 1);
  }

  render() {
    const {
      sorts,
      pageSize,
      page,
    } = this.state;
    const {
      /* eslint-disable no-unused-vars */
      initialSorts,
      pageSize: pageSizeProp,
      onChange,
      /* eslint-enable no-unused-vars */
      data,
      columns,
      loading,
      count,
      showPagination,
      showPageJump,
      showPageSizeOptions,
      pageSizeOptions,
      TableComponent,
      TheadComponent,
      TbodyComponent,
      TrComponent,
      ThComponent,
      TdComponent,
      PaginationComponent,
      PreviousComponent,
      NextComponent,
      LoadingComponent,
      tableStyle,
      theadStyle,
      tbodyStyle,
      trStyle,
      thStyle,
      tdStyle,
      paginationStyle,
      previousStyle,
      nextStyle,
      loadingStyle,
      tableClassName,
      theadClassName,
      tbodyClassName,
      trClassName,
      thClassName,
      tdClassName,
      paginationClassName,
      previousClassName,
      nextClassName,
      loadingClassName,
      ...rest
    } = this.props;
    return (
      <div {...rest}>
        { loading ? <LoadingComponent style={loadingStyle} className={loadingClassName} /> : null }
        <TableComponent
          style={tableStyle}
          className={tableClassName}
          data={data}
          columns={columns}
          sorts={sorts}
          handleHeaderClick={this.changeSort}
          TheadComponent={TheadComponent}
          TbodyComponent={TbodyComponent}
          TrComponent={TrComponent}
          ThComponent={ThComponent}
          TdComponent={TdComponent}
          theadStyle={theadStyle}
          tbodyStyle={tbodyStyle}
          trStyle={trStyle}
          thStyle={thStyle}
          tdStyle={tdStyle}
          theadClassName={theadClassName}
          tbodyClassName={tbodyClassName}
          trClassName={trClassName}
          thClassName={thClassName}
          tdClassName={tdClassName}
        />
        {
          showPagination ? (
            <PaginationComponent
              style={paginationStyle}
              className={paginationClassName}
              page={page}
              count={count}
              pageSize={pageSize}
              showPageJump={showPageJump}
              showPageSizeOptions={showPageSizeOptions}
              pageSizeOptions={pageSizeOptions}
              PreviousComponent={PreviousComponent}
              NextComponent={NextComponent}
              previousStyle={previousStyle}
              nextStyle={nextStyle}
              previousClassName={previousClassName}
              nextClassName={nextClassName}
              setPageSize={this.setPageSize}
              jumpToPage={this.jumpToPage}
              incrementPage={this.incrementPage}
              decrementPage={this.decrementPage}
            />
          ) : null
        }
      </div>
    );
  }
}

ReactTable.defaultProps = {
  loading: false,
  pageSize: 20,
  showPagination: true,
  showPageJump: true,
  showPageSizeOptions: true,
  pageSizeOptions: [5, 10, 20, 25, 50, 100],
  TableComponent,
  TheadComponent,
  TbodyComponent,
  TrComponent,
  ThComponent,
  TdComponent,
  PaginationComponent,
  PreviousComponent,
  NextComponent,
  LoadingComponent,
  style: {},
  tableStyle: {},
  theadStyle: {},
  tbodyStyle: {},
  trStyle: {},
  thStyle: {},
  tdStyle: {},
  paginationStyle: {},
  previousStyle: {},
  nextStyle: {},
  loadingStyle: {},
  className: '',
  tableClassName: '',
  theadClassName: '',
  tbodyClassName: '',
  trClassName: '',
  thClassName: '',
  tdClassName: '',
  paginationClassName: '',
  previousClassName: '',
  nextClassName: '',
  loadingClassName: '',
};

ReactTable.propTypes = {
  data: dataType.isRequired,
  columns: PropTypes.arrayOf(columnType).isRequired,
  initialSorts: PropTypes.arrayOf(sortType).isRequired,
  count: PropTypes.number.isRequired,
  onChange: PropTypes.func.isRequired,
  pageSize: PropTypes.number,
  loading: PropTypes.bool,
  showPagination: PropTypes.bool,
  showPageJump: PropTypes.bool,
  showPageSizeOptions: PropTypes.bool,
  pageSizeOptions: PropTypes.arrayOf(PropTypes.number),
  TableComponent: PropTypes.func,
  TheadComponent: PropTypes.func,
  TbodyComponent: PropTypes.func,
  TrComponent: PropTypes.func,
  ThComponent: PropTypes.func,
  TdComponent: PropTypes.func,
  PaginationComponent: PropTypes.func,
  PreviousComponent: PropTypes.func,
  NextComponent: PropTypes.func,
  LoadingComponent: PropTypes.func,
  style: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  tableStyle: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  theadStyle: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  tbodyStyle: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  trStyle: styleType,
  thStyle: styleType,
  tdStyle: styleType,
  paginationStyle: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  previousStyle: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  nextStyle: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  loadingStyle: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  className: PropTypes.string,
  tableClassName: PropTypes.string,
  theadClassName: PropTypes.string,
  tbodyClassName: PropTypes.string,
  trClassName: classNameType,
  thClassName: classNameType,
  tdClassName: classNameType,
  paginationClassName: PropTypes.string,
  previousClassName: PropTypes.string,
  nextClassName: PropTypes.string,
  loadingClassName: PropTypes.string,
};

export default ReactTable;
