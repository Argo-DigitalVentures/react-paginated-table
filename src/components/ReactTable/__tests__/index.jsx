import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import ReactTable from '../';

const data = [
  { company: 'Argo', location: 'New York' },
  { company: 'John Hancock', location: 'New York' },
];

const columns = [
  { id: 'company', header: 'Company Name', selector: 'company' },
  { id: 'location', header: () => null, selector: 'location' },
];

describe('<ReactTable />', () => {
  it('should use custom style, className, and other DOM props', () => {
    const tree = shallow(
      <ReactTable
        style={{
          color: 'green',
        }}
        className="test"
        data={data}
        columns={columns}
        initialSorts={[]}
        count={10}
        onChange={() => null}
      />,
    );
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should initialize sorts and pageSize state based on props', () => {
    const tree = shallow(
      <ReactTable
        style={{
          color: 'green',
        }}
        className="test"
        data={data}
        columns={columns}
        initialSorts={[{ columnId: 'test', direction: 'DESC' }]}
        pageSize={42}
        count={10}
        onChange={() => null}
      />,
    );
    expect(tree.instance().state.sorts).toEqual([{ columnId: 'test', direction: 'DESC' }]);
    expect(tree.instance().state.pageSize).toEqual(42);
  });

  it('should display the loading component if loading prop is true', () => {
    const tree = shallow(
      <ReactTable
        style={{
          color: 'green',
        }}
        className="test"
        data={data}
        columns={columns}
        initialSorts={[]}
        count={10}
        loading
        onChange={() => null}
      />,
    );
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should hide the pagination component if showPagination prop is false', () => {
    const tree = shallow(
      <ReactTable
        style={{
          color: 'green',
        }}
        className="test"
        data={data}
        columns={columns}
        initialSorts={[]}
        count={10}
        showPagination={false}
        onChange={() => null}
      />,
    );
    expect(toJson(tree)).toMatchSnapshot();
  });

  describe('onChange()', () => {
    it('should be invoked on mount and call the onChange prop when invoked', () => {
      const onChangeSpy = jest.fn();
      const tree = shallow(
        <ReactTable
          style={{
            color: 'green',
          }}
          className="test"
          data={data}
          columns={columns}
          initialSorts={[{ columnId: 'test', direction: 'DESC' }]}
          count={10}
          onChange={onChangeSpy}
        />,
      );
      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      tree.instance().onChange();
      expect(onChangeSpy).toHaveBeenCalledTimes(2);
      expect(onChangeSpy).toBeCalledWith({
        sorts: [{ columnId: 'test', direction: 'DESC' }],
        pageSize: 20,
        page: 1,
        count: 10,
      });
    });
  });

  describe('changeSort()', () => {
    it('should add, replace, and update sorts', () => {
      const onChangeSpy = jest.fn();
      const tree = shallow(
        <ReactTable
          style={{
            color: 'green',
          }}
          className="test"
          data={data}
          columns={columns}
          initialSorts={[{ columnId: 'test', direction: 'DESC' }]}
          count={10}
          onChange={onChangeSpy}
        />,
      );

      expect(onChangeSpy).toHaveBeenCalledTimes(1);

      tree.instance().changeSort('other', true);
      expect(tree.instance().state.sorts).toEqual([
        { columnId: 'test', direction: 'DESC' },
        { columnId: 'other', direction: 'ASC' },
      ]);
      expect(onChangeSpy).toHaveBeenCalledTimes(2);

      tree.instance().changeSort('test', true);
      expect(tree.instance().state.sorts).toEqual([
        { columnId: 'other', direction: 'ASC' },
      ]);
      expect(onChangeSpy).toHaveBeenCalledTimes(3);

      tree.instance().changeSort('another', false);
      expect(tree.instance().state.sorts).toEqual([
        { columnId: 'another', direction: 'ASC' },
      ]);
      expect(onChangeSpy).toHaveBeenCalledTimes(4);

      tree.instance().changeSort('another', false);
      expect(tree.instance().state.sorts).toEqual([
        { columnId: 'another', direction: 'DESC' },
      ]);
      expect(onChangeSpy).toHaveBeenCalledTimes(5);

      tree.instance().changeSort('another', false);
      expect(tree.instance().state.sorts).toEqual([]);
      expect(onChangeSpy).toHaveBeenCalledTimes(6);
    });
  });

  describe('jumpToPage()', () => {
    it('should change the page and call the onChange prop', () => {
      const onChangeSpy = jest.fn();
      const tree = shallow(
        <ReactTable
          style={{
            color: 'green',
          }}
          className="test"
          data={data}
          columns={columns}
          initialSorts={[]}
          count={100}
          onChange={onChangeSpy}
        />,
      );
      const instance = tree.instance();
      expect(instance.state.page).toBe(1);
      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      instance.jumpToPage(3);
      expect(instance.state.page).toBe(3);
      expect(onChangeSpy).toHaveBeenCalledTimes(2);
    });
  });

  describe('incrementPage()', () => {
    it('should increment the page and call the onChange prop', () => {
      const onChangeSpy = jest.fn();
      const tree = shallow(
        <ReactTable
          style={{
            color: 'green',
          }}
          className="test"
          data={data}
          columns={columns}
          initialSorts={[]}
          count={100}
          onChange={onChangeSpy}
        />,
      );
      const instance = tree.instance();
      expect(instance.state.page).toBe(1);
      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      instance.incrementPage();
      expect(instance.state.page).toBe(2);
      expect(onChangeSpy).toHaveBeenCalledTimes(2);
      instance.incrementPage();
      expect(instance.state.page).toBe(3);
      expect(onChangeSpy).toHaveBeenCalledTimes(3);
    });
  });

  describe('decrementPage()', () => {
    it('should decrement the page and call the onChange prop', () => {
      const onChangeSpy = jest.fn();
      const tree = shallow(
        <ReactTable
          style={{
            color: 'green',
          }}
          className="test"
          data={data}
          columns={columns}
          initialSorts={[]}
          count={100}
          onChange={onChangeSpy}
        />,
      );
      const instance = tree.instance();
      expect(instance.state.page).toBe(1);
      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      instance.jumpToPage(5);
      expect(instance.state.page).toBe(5);
      expect(onChangeSpy).toHaveBeenCalledTimes(2);
      instance.decrementPage();
      expect(instance.state.page).toBe(4);
      expect(onChangeSpy).toHaveBeenCalledTimes(3);
      instance.decrementPage();
      expect(instance.state.page).toBe(3);
      expect(onChangeSpy).toHaveBeenCalledTimes(4);
    });
  });

  describe('setPageSize()', () => {
    it('should decrement the page and call the onChange prop', () => {
      const onChangeSpy = jest.fn();
      const tree = shallow(
        <ReactTable
          style={{
            color: 'green',
          }}
          className="test"
          data={data}
          columns={columns}
          initialSorts={[]}
          count={100}
          onChange={onChangeSpy}
        />,
      );
      const instance = tree.instance();
      expect(instance.state.pageSize).toBe(20);
      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      instance.setPageSize(50);
      expect(instance.state.pageSize).toBe(50);
      expect(onChangeSpy).toHaveBeenCalledTimes(2);
      instance.setPageSize(100);
      expect(instance.state.pageSize).toBe(100);
      expect(onChangeSpy).toHaveBeenCalledTimes(3);
    });
  });
  // setPageSize
});
