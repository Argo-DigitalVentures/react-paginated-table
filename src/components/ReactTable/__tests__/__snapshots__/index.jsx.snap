exports[`<ReactTable /> should display the loading component if loading prop is true 1`] = `
<div
  className="test"
  style={
    Object {
      "color": "green",
    }
  }>
  <Loading
    className=""
    style={Object {}} />
  <Table
    TbodyComponent={[Function]}
    TdComponent={[Function]}
    ThComponent={[Function]}
    TheadComponent={[Function]}
    TrComponent={[Function]}
    className=""
    columns={
      Array [
        Object {
          "header": "Company Name",
          "id": "company",
          "selector": "company",
        },
        Object {
          "header": [Function],
          "id": "location",
          "selector": "location",
        },
      ]
    }
    data={
      Array [
        Object {
          "company": "Argo",
          "location": "New York",
        },
        Object {
          "company": "John Hancock",
          "location": "New York",
        },
      ]
    }
    handleHeaderClick={[Function]}
    sorts={Array []}
    style={Object {}}
    tbodyClassName=""
    tbodyStyle={Object {}}
    tdClassName=""
    tdStyle={Object {}}
    thClassName=""
    thStyle={Object {}}
    theadClassName=""
    theadStyle={Object {}}
    trClassName=""
    trStyle={Object {}} />
  <Pagination
    NextComponent={[Function]}
    PreviousComponent={[Function]}
    className=""
    count={10}
    decrementPage={[Function]}
    incrementPage={[Function]}
    jumpToPage={[Function]}
    nextClassName=""
    nextStyle={Object {}}
    page={1}
    pageSize={20}
    pageSizeOptions={
      Array [
        5,
        10,
        20,
        25,
        50,
        100,
      ]
    }
    previousClassName=""
    previousStyle={Object {}}
    setPageSize={[Function]}
    showPageJump={true}
    showPageSizeOptions={true}
    style={Object {}} />
</div>
`;

exports[`<ReactTable /> should hide the pagination component if showPagination prop is false 1`] = `
<div
  className="test"
  style={
    Object {
      "color": "green",
    }
  }>
  <Table
    TbodyComponent={[Function]}
    TdComponent={[Function]}
    ThComponent={[Function]}
    TheadComponent={[Function]}
    TrComponent={[Function]}
    className=""
    columns={
      Array [
        Object {
          "header": "Company Name",
          "id": "company",
          "selector": "company",
        },
        Object {
          "header": [Function],
          "id": "location",
          "selector": "location",
        },
      ]
    }
    data={
      Array [
        Object {
          "company": "Argo",
          "location": "New York",
        },
        Object {
          "company": "John Hancock",
          "location": "New York",
        },
      ]
    }
    handleHeaderClick={[Function]}
    sorts={Array []}
    style={Object {}}
    tbodyClassName=""
    tbodyStyle={Object {}}
    tdClassName=""
    tdStyle={Object {}}
    thClassName=""
    thStyle={Object {}}
    theadClassName=""
    theadStyle={Object {}}
    trClassName=""
    trStyle={Object {}} />
</div>
`;

exports[`<ReactTable /> should use custom style, className, and other DOM props 1`] = `
<div
  className="test"
  style={
    Object {
      "color": "green",
    }
  }>
  <Table
    TbodyComponent={[Function]}
    TdComponent={[Function]}
    ThComponent={[Function]}
    TheadComponent={[Function]}
    TrComponent={[Function]}
    className=""
    columns={
      Array [
        Object {
          "header": "Company Name",
          "id": "company",
          "selector": "company",
        },
        Object {
          "header": [Function],
          "id": "location",
          "selector": "location",
        },
      ]
    }
    data={
      Array [
        Object {
          "company": "Argo",
          "location": "New York",
        },
        Object {
          "company": "John Hancock",
          "location": "New York",
        },
      ]
    }
    handleHeaderClick={[Function]}
    sorts={Array []}
    style={Object {}}
    tbodyClassName=""
    tbodyStyle={Object {}}
    tdClassName=""
    tdStyle={Object {}}
    thClassName=""
    thStyle={Object {}}
    theadClassName=""
    theadStyle={Object {}}
    trClassName=""
    trStyle={Object {}} />
  <Pagination
    NextComponent={[Function]}
    PreviousComponent={[Function]}
    className=""
    count={10}
    decrementPage={[Function]}
    incrementPage={[Function]}
    jumpToPage={[Function]}
    nextClassName=""
    nextStyle={Object {}}
    page={1}
    pageSize={20}
    pageSizeOptions={
      Array [
        5,
        10,
        20,
        25,
        50,
        100,
      ]
    }
    previousClassName=""
    previousStyle={Object {}}
    setPageSize={[Function]}
    showPageJump={true}
    showPageSizeOptions={true}
    style={Object {}} />
</div>
`;
